20240627-BC-listapi.md

```diff
@@ -0,0 +1,182 @@
+
+## **Sorting**
+
+Sorting is required for token-based pagination in relational databases, so we must have a default sorting order.
+
+Some no-sql databases (ex. Azure blob store) don’t support sorting in descending order and others don’t support any sorting at all (ex. Redis). In these cases, we want to return an explicit error instead of failing silently.
+
+This might be restricting for use cases where the underlying state store needs to be swapped though. For example a team could use Redis for local development, and Postgres in production, and they wouldn’t be able to use the same application code, because the sorting clause would error on Redis, but pass on Postgres. That’s why we’re introducing the `Default` sorting option which will sort in ascending order for all databases that support it, and leave results unsorted for the databases that don’t.
+
+## SDKs
+
+All supported SDKs should be updated to implement the List API. SDKs should offer the option to fetch batch values of the returned keys.
+
+## Default behaviour for state stores with missing features
+
+Some of the state stores Dapr supports don’t provide the necessary capabilities for implementing the list API. For example, Memcached doesn’t provide a way to list keys,  Azure table storage can’t sort keys in descending order and so on. For those cases the list API will do a best effort to provide the closest functionality to the one defined in the API. The functionality will be specific to the data store and will be implemented on the component level.
+
+List API requests on state stores that don’t support the List API will result in errors.
+
+## Impact of the List API on Dapr state store components
+From the moment this proposal is accepted, all state store components will be required to implement the List API in order to get the "Stable" certification level.
+Components that are currently stable and for which the underlying state store does not support listing will not lose their stable status.
+
+## Performance and pricing implications
+Listing keys in big data sets, specially for partitioned databases, can be expensive in terms of both performance and cost. Often it would incur creating an index which will impact write performance, storage cost and sometimes even read performance.
+For the databases where this is a concern, we should offer an option to disable the List API on the component level.
+
+## Definitions:
+
+- **Listing**: The ability to retrieve a collection of items.
+- **Sorting**: The ability to sort the results based on one or more fields.
+- **Prefix Search**: The ability to search for items that start with a given prefix.
+- **Pagination**: The ability to paginate through items, typically using skip/limit or similar mechanisms.
+
+## Adendum
+
+Here's a list of the relevant capabilities of all the stable state stores:
+
+| Store | Cursor listing | Offset listing | Sorting | Number of Items per Page | Prefix Search | Comments |
```

Comment:
From this list, we can clearly see that with cursor listing, page limit and prefix search, we will have plenty of coverage.